File: lib/core/constants/humanization_pass.dart
import '../utils/payload_optimization.dart';
import 'audio_environment_data.dart';
import 'dialect_style_data.dart';
import 'vocal_accent_data.dart';

/// Runtime post-processing: humanization pass (OpenRouter + LaoZhang).
///
/// Source: tools/humanization_pass.txt — keep [systemPrompt] aligned with that
/// file; [buildUserMessage] is hand-maintained Dart-only builder logic.
abstract final class HumanizationPassConfig {
  HumanizationPassConfig._();

  static const String systemPrompt = '''

# POST-PROCESSING: HUMANIZATION PASS (RUNTIME — BLOCK 2 LYRICS)

**Model (runtime):**
- **OpenRouter:** Mistral Large — rewrite lane (natural phrasing, anti-AI, authenticity).
- **LaoZhang English song context:** Claude Sonnet 4.5 — standard English lyrics, accent-vs-dialect rules.
- **LaoZhang multilingual / Nigerian Pidgin / African languages:** GPT-5.5 — honor Language field and Pidgin grammar; never flatten to textbook English.

**When:** After theme consistency, before Suno compression (OpenRouter + LaoZhang). **Silent** — no commentary.

**Scope:** Revise **performable lyric lines** in Block 2. Keep section headers, arrangement staging lines, performance tags, and `[End]` unchanged.

---

## OBJECTIVE

Humanize lyrics: natural speech, genre-appropriate authenticity, zero AI-artifact voice.

---

## ACTIONS

1. **Humanize** — conversational, singable, not essay or caption voice.
2. **Natural phrasing** — how a real artist would say it on a record.
3. **Remove AI artifacts** — ban poster slogans, motivational copy, generic philosophy, symmetrical "polished" chains.
4. **Improve authenticity** — apply genre lane from user GENRE (Amapiano conversational · Hip-hop attitude · Country place-based · Trance singable hooks · Gospel spirit-forward · etc.).
5. **Nigerian Pidgin (when user LYRIC DIALECT says so):** Write natively in authentic Pidgin — `dey`, `na`, `wahala`, `e don set`, `small small`. Do **not** normalize to standard English. Fourth-Wall Law still applies in lyric lines.
6. **Multilingual / African languages (when user Language is not English):** Preserve native lyric intent, grammar, and cultural phrasing — do not translate into stiff English.

---

## ACCENT VS. DIALECT CONSTRAINT (STAGE 3 — MANDATORY WHEN ACCENT IS SET)

When a regional accent or delivery style is specified (e.g. Nigerian, British, Jamaican) **and** lyric dialect is **not** Nigerian Pidgin:

* Write lyrics in **clear, standard English**.
* You are **strictly forbidden** from translating text into slang, broken dialects, or patois (completely ban words like `dey`, `na`, `wahala`, `gonna` in lyric lines).
* Vocal performance style stays **purely phonetic** in staging tags; written lyric text must stay pristine, high-end, and universally legible.

**Exception:** When user LYRIC DIALECT = Nigerian Pidgin, Pidgin grammar in lyric lines is required — accent still lives in staging tags only.

---

## PRESERVE

- Syllable feel · groove · rhyme · cadence · hook anchors (unless AI-sloppy)
- All bracket structure (headers → staging → tags → lines → `[End]`)

---

## OUTPUT

Return **only** the revised Block 2 body. No Block 1. No scores. No meta.

''';

  static String buildUserMessage({
    required String block2Body,
    required String primaryGenre,
    required String subGenreFusion,
    required String vibe,
    required String lyricThemeNotes,
    required String language,
    String? vocalAccent,
    String? dialectStyleId,
    String? dialectVariantId,
    String? audioEnvironmentModeId,
  }) {
    final resolvedDialectId =
        dialectStyleId ?? DialectStyleData.standardEnglishId;

    final extraLines = _buildExtraLines(
      vocalAccent: vocalAccent,
      resolvedDialectId: resolvedDialectId,
      dialectStyleId: dialectStyleId,
      dialectVariantId: dialectVariantId,
      audioEnvironmentModeId: audioEnvironmentModeId,
    );

    final contextHeader = compactGenreContextHeader(
      primaryGenre: primaryGenre,
      subGenreFusion: subGenreFusion,
      vibe: vibe,
      lyricThemeNotes: lyricThemeNotes,
      language: language,
      extraLines: extraLines,
    );

    final compactBody = compactPayloadText(block2Body);

    return '''
$contextHeader

BLOCK 2 (humanize; keep structure):
---
$compactBody
---

Return ONLY revised Block 2 through [End]. No Block 1. No commentary.
'''.trim();
  }

  static List<String> _buildExtraLines({
    required String? vocalAccent,
    required String resolvedDialectId,
    required String? dialectStyleId,
    required String? dialectVariantId,
    required String? audioEnvironmentModeId,
  }) {
    // Compact helpers return '' when inactive; filter empties so the header
    // stays dense. If any helper becomes nullable later, coalesce before filter.
    return <String>[
      VocalAccentData.postProcessCompactLine(
        vocalAccent,
        dialectStyleId: resolvedDialectId,
      ),
      DialectStyleData.postProcessCompactLine(
        dialectStyleId,
        dialectVariantId: dialectVariantId,
      ),
      VocalAccentData.accentVsDialectCompactLine(
        vocalAccent,
        dialectStyleId: resolvedDialectId,
      ),
      AudioEnvironmentData.postProcessCompactLine(audioEnvironmentModeId),
    ].where((line) => line.isNotEmpty).toList(growable: false);
  }
}

// Backward-compatible top-level names.
const String kHumanizationSystemPrompt = HumanizationPassConfig.systemPrompt;

String buildHumanizationUserMessage({
  required String block2Body,
  required String primaryGenre,
  required String subGenreFusion,
  required String vibe,
  required String lyricThemeNotes,
  required String language,
  String? vocalAccent,
  String? dialectStyleId,
  String? dialectVariantId,
  String? audioEnvironmentModeId,
}) =>
    HumanizationPassConfig.buildUserMessage(
      block2Body: block2Body,
      primaryGenre: primaryGenre,
      subGenreFusion: subGenreFusion,
      vibe: vibe,
      lyricThemeNotes: lyricThemeNotes,
      language: language,
      vocalAccent: vocalAccent,
      dialectStyleId: dialectStyleId,
      dialectVariantId: dialectVariantId,
      audioEnvironmentModeId: audioEnvironmentModeId,
    );

File: server/app/humanization_pass.py
"""Post-processing: humanization pass on Block 2 (OpenRouter + LaoZhang)."""

from __future__ import annotations

import logging
from typing import Any

from app.llm_config import llm_provider, resolve_humanization_model
from app.payload_optimization import compact_payload_text
from app.post_process_common import (
    block2_output_valid,
    chat_simple,
    genre_context_block,
    load_tool_prompt,
    merge_block2_parts,
    sanitize_block2_output,
    split_block2_parts,
    truthy_env,
)
from app.suno_output_qa import unified_block2_missing

logger = logging.getLogger(__name__)

_FALLBACK = """HUMANIZATION PASS — Block 2 lyrics only.
Humanize, natural phrasing, remove AI artifacts, genre authenticity.
Preserve headers, staging, tags, syllables, hooks, [End]. Output Block 2 body only."""


def humanization_pass_enabled(*, provider: str | None = None) -> bool:
    if not truthy_env("HUMANIZATION_PASS", default=True):
        return False
    return llm_provider(provider) in ("openrouter", "laozhang")


def apply_humanization_pass(
    client: Any,
    *,
    text: str,
    primary_genre: str = "",
    sub_genre_fusion: str = "",
    vibe: str = "",
    lyric_theme_notes: str = "",
    language: str = "English",
    vocal_accent: str = "",
    dialect_style_id: str = "",
    dialect_variant_id: str = "",
    audio_environment_mode: str = "",
    provider: str | None = None,
    lyrics_task: bool = True,
    remix_original_song_title: str = "",
    remix_original_artist: str = "",
    song_generation_type: str = "full_song",
) -> tuple[str, str]:
    if not humanization_pass_enabled(provider=provider) or not lyrics_task:
        return text, "humanize-skip"
    if unified_block2_missing(text):
        return text, "humanize-skip"

    parts = split_block2_parts(text)
    if parts is None:
        return text, "humanize-skip"
    prefix, block2_body, suffix = parts
    if not block2_body.strip():
        return text, "humanize-skip"

    system = load_tool_prompt("humanization_pass.txt", fallback=_FALLBACK)
    ctx = genre_context_block(
        primary_genre=primary_genre,
        sub_genre_fusion=sub_genre_fusion,
        vibe=vibe,
        lyric_theme_notes=lyric_theme_notes,
        language=language,
        vocal_accent=vocal_accent,
        dialect_style_id=dialect_style_id,
        dialect_variant_id=dialect_variant_id,
        audio_environment_mode=audio_environment_mode,
        remix_original_song_title=remix_original_song_title,
        remix_original_artist=remix_original_artist,
        song_generation_type=song_generation_type,
    )
    body = compact_payload_text(block2_body)
    user_msg = (
        f"{ctx}\n\nBLOCK 2 (humanize; keep structure):\n---\n{body}\n---\n"
        "Return ONLY revised Block 2 through [End]. No Block 1. No commentary."
    )

    model = resolve_humanization_model(
        provider=provider,
        language=language,
        dialect_style_id=dialect_style_id,
    )
    try:
        logger.info("humanization pass model=%s", model)
        out = chat_simple(
            client,
            model=model,
            system=system,
            user=user_msg,
            temperature=0.45,
        )
        candidate = sanitize_block2_output(out)
        if not block2_output_valid(candidate):
            return text, "humanize-invalid"
        if candidate == block2_body:
            return text, "humanize-unchanged"
        merged = merge_block2_parts(prefix, candidate, suffix)
        if unified_block2_missing(merged):
            return text, "humanize-fail"
        return merged, f"humanize:{model}"
    except Exception as exc:  # noqa: BLE001
        logger.warning("humanization pass failed: %s", exc)
        return text, "humanize-fail"
